lib/_http_client.js

```diff
@@ -562,7 +562,13 @@ function tickOnSocket(req, socket) {
   socket.on('close', socketCloseListener);
 
   if (req.timeout) {
-    socket.once('timeout', () => req.emit('timeout'));
+    const emitRequestTimeout = () => req.emit('timeout');
```

Comment:
I reused your unit test code and took a stab at the approach described above in #9831 

The primary difference is that any timeout set applies from the time a socket is assigned to a request, i.e. it includes socket pre-connect timeouts.  In existing code, if a timeout is set via `ClientRequest.prototype.setTimeout`, it is "reapplied" and starts counting only from the time the socket connects.  Perhaps this would count as an API change.

In other HTTP libraries, the connect-timeout and the transaction-timeout are two separate parameters, e.g. in curl https://curl.haxx.se/libcurl/c/CURLOPT_CONNECTTIMEOUT.html and https://curl.haxx.se/libcurl/c/CURLOPT_TIMEOUT.html